batch_job/src/batch_wrapper.c

```diff
@@ -0,0 +1,324 @@
+	unsigned i = 0;
+	for (void *x = NULL; (j = jx_iterate_array(argv, &x));) {
+		if (jx_istype(j, JX_STRING)) {
+			ptrs[i] = xxstrdup(j->u.string_value);
+		} else if (jx_istype(j, JX_OBJECT)) {
+			ptrs[i] = batch_wrapper_expand(t, j);
+			if (!ptrs[i]) {
+				free_argv(ptrs);
+				return NULL;
+			}
+		} else {
+			free_argv(ptrs);
+			debug(D_NOTICE|D_BATCH, "arguments must be strings");
+			return NULL;
+		}
+		++i;
+	}
+
+	return ptrs;
+}
+
+char *batch_wrapper_expand(struct batch_task *t, struct jx *spec) {
+	assert(t);
+	assert(spec);
+
+	struct jx *j;
+	unsigned commands = 0;
+	char *result = NULL;
+	struct batch_wrapper *w = batch_wrapper_create();
+
+	if (!jx_istype(spec, JX_OBJECT)) {
+		debug(D_NOTICE|D_BATCH, "wrapper command spec must be a JX object");
+		goto FAIL;
+	}
+
+	struct jx *pre = jx_lookup(spec, "pre");
+	if (pre) {
+		if (!jx_istype(pre, JX_ARRAY)) {
+			debug(D_NOTICE|D_BATCH, "pre commands must be specified in an array");
+			goto FAIL;
+		}
+		for (void *i = NULL; (j = jx_iterate_array(pre, &i));) {
+			if (!jx_istype(j, JX_STRING)) {
+				debug(D_NOTICE|D_BATCH, "pre commands must be strings");
+				goto FAIL;
+			}
+			batch_wrapper_pre(w, j->u.string_value);
+		}
+	}
+
+	struct jx *post = jx_lookup(spec, "post");
+	if (post) {
+		if (!jx_istype(post, JX_ARRAY)) {
+			debug(D_NOTICE|D_BATCH, "post commands must be specified in an array");
+			goto FAIL;
+		}
+		for (void *i = NULL; (j = jx_iterate_array(post, &i));) {
+			if (!jx_istype(j, JX_STRING)) {
+				debug(D_NOTICE|D_BATCH, "post commands must be strings");
+				goto FAIL;
+			}
+			batch_wrapper_post(w, j->u.string_value);
+		}
+	}
+
+	struct jx *argv = jx_lookup(spec, "argv");
+	if (argv) {
+		if (commands++) {
+			debug(D_NOTICE|D_BATCH, "only one command is allowed");
+			goto FAIL;
+		}
+
+		char **ptrs = jx_array_to_argv(t, argv);
+		if (ptrs) {
+			batch_wrapper_argv(w, ptrs);
+			free_argv(ptrs);
+		} else {
+			free_argv(ptrs);
+			goto FAIL;
+		}
+	}
+
+	struct jx *args = jx_lookup(spec, "args");
+	if (args) {
+		if (commands++) {
+			debug(D_NOTICE|D_BATCH, "only one command is allowed");
+			goto FAIL;
+		}
+
+		char **ptrs = jx_array_to_argv(t, args);
+		if (ptrs) {
+			batch_wrapper_args(w, ptrs);
+			free_argv(ptrs);
+		} else {
+			free_argv(ptrs);
+			goto FAIL;
+		}
+	}
+
+	struct jx *cmd = jx_lookup(spec, "cmd");
+	if (cmd) {
+		if (commands++) {
+			debug(D_NOTICE|D_BATCH, "only one command is allowed");
+			goto FAIL;
+		}
+
+		if (jx_istype(cmd, JX_OBJECT)) {
+			char *nested = batch_wrapper_expand(t, cmd);
+			if (!nested) goto FAIL;
+			batch_wrapper_cmd(w, nested);
+			free(nested);
+		} else if (jx_istype(cmd, JX_STRING)) {
+			batch_wrapper_cmd(w, cmd->u.string_value);
+		} else {
+			debug(D_NOTICE|D_BATCH, "cmd must be a string");
+			goto FAIL;
+		}
+	}
+
+	if (commands != 1) {
+		debug(D_NOTICE|D_BATCH, "a command is required to generate a wrapper");
+		goto FAIL;
+	}
+
+	result = batch_wrapper_write(w, t, "./wrapper_");
```

Comment:
Can we parse out a wrapper name instead of just using wrapper? I think wrapper is an adequate default, but would prefer being able to name it something more descriptive if possible.